### PR TITLE
refactoring registration service to make it resilient for conflict errors

### PR DESCRIPTION
**Pull request by abhupadh**
## Description

See [issue](https://github.com/adobe/aio-lib-java/issues/199) description

## Related Issue
https://github.com/adobe/aio-lib-java/issues/199
https://jira.corp.adobe.com/browse/CI-6365

## Motivation and Context

Make our e2e more resilient

## How Has This Been Tested?

Using `aio-events-itests` running locally

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the [code style](CODE_STYLE.md) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

**Review by francoisledroff — COMMENTED** [review]
cool, thanks for the PR, it looks good to me
as you are at it, could you also add the same use of the `x-conflicting-id` header in `ProviderService.createOrUpdateProvider(..)` ?

**Comment by abhupadh** [comment]
@francoisledroff : please approve

File: events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java [hunk 1]
```diff
@@ -51,7 +51,7 @@ public void createGetDeleteJournalRegistration() {
     try {
       providerId = providerServiceTester.createOrUpdateProvider(TEST_EVENT_PROVIDER_LABEL,
           TEST_EVENT_CODE).getId();
-      Registration registration = createJournalRegistration(TEST_REGISTRATION_NAME, providerId,
+      Registration registration = createOrUpdateJournalRegistration(TEST_REGISTRATION_NAME, providerId,
```
Comment (francoisledroff) [inline events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java:54]: may be we could test both path ?
using `createOrUpdate` once and then `createOrUpdate` again ... once created 
here and may be in `ProviderServiceIntegrationTest` too 
wdyt ?

**Review by francoisledroff — CHANGES_REQUESTED** [review]
a few small requests:

File: events_mgmt/src/main/java/com/adobe/aio/event/management/feign/FeignRegistrationService.java [hunk 2]
```diff
@@ -70,6 +73,18 @@ public void delete(String registrationId) {
...
   }
 
+  @Override
+  public Optional<Registration> createOrUpdateRegistration(
+      RegistrationCreateModel.Builder registrationCreateModelBuilder) {
+    try {
+      return createRegistration(registrationCreateModelBuilder);
+    } catch (ConflictException ex) {
+      String conflictingRegistrationId = ex.getConflictingId();
+      logger.warn("Another registration (id `{}`) exists with conflict due to {}. Trying to update it...", conflictingRegistrationId, ex.getMessage());
```
[the remark this answers is not in the document]
Comment (abhupadh) [reply events_mgmt/src/main/java/com/adobe/aio/event/management/feign/FeignRegistrationService.java:83]: ok

File: events_mgmt/src/main/java/com/adobe/aio/event/management/feign/FeignProviderService.java [hunk 3]
```diff
@@ -108,22 +108,11 @@ public Optional<Provider> createProvider(final ProviderInputModel providerInputM
...
-              + "(`" + providerInputModel.getProviderMetadataId() + "`,"
-              + "`" + workspace.getImsOrgId() + "`,"
-              + "`" + providerInputModel.getInstanceId()+ "`)"
-              + " may have been deleted just after a Conflict `" + e.getMessage()
-              + "` was detected while creating it"))
-          .getId();
-      return updateProvider(providerId,providerInputModel);
+    } catch (ConflictException e){
+      String conflictingProviderId = e.getConflictingId();
+      logger.warn("Another provider (id: `{}` ) exist with conflict due to {}, trying to update it ...",
```
[the remark this answers is not in the document]
Comment (abhupadh) [reply events_mgmt/src/main/java/com/adobe/aio/event/management/feign/FeignProviderService.java:113]: ok

File: events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java (hunk 1 above)
Comment (abhupadh) [reply events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java:54]: I didn't get this

**Comment by abhupadh** [comment]
@francoisledroff : how can we retrigger the integration test? Locally it is getting passed.

File: events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java (hunk 1 above)
Comment (abhupadh) [reply events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java:55]: done

File: events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java [hunk 4]
```diff
@@ -51,24 +52,19 @@ public void createGetDeleteJournalRegistration() {
...
-      assertEquals(registration.isEnabled(), found.get().isEnabled());
-      assertEquals(registration.getWebhookUrl(), found.get().getWebhookUrl());
-      assertEquals(registration.getJournalUrl().getHref(), found.get().getJournalUrl().getHref());
-      assertEquals(registration.getTraceUrl().getHref(), found.get().getTraceUrl().getHref());
+      String finalRegistrationId = registration.getRegistrationId();
+      assertCreatedOrUpdatedRegistrationMatchesWithFoundRegistration(registration,
+          () -> registrationService.findById(finalRegistrationId));
+
+      // covering the update path
+      registration = createOrUpdateJournalRegistration(TEST_REGISTRATION_NAME, providerId,
```
Comment (francoisledroff) [inline events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java:63]: could you add the same coverage for the `ProviderService` pretty please ?

**Review by francoisledroff — CHANGES_REQUESTED** [review]
one small request:

File: events_test/src/test/java/com/adobe/aio/event/management/ProviderServiceIntegrationTest.java [hunk 5]
```diff
@@ -107,34 +109,29 @@ public void createGetUpdateDelete() {
...
-      } catch (ConflictException ex) {
-        logger.info("Cannot create an AIO Events provider with the same instanceId: {}",
-            ex.getMessage());
-      }
-
+      // update provider and event metadata
       String updatedProviderDescription = "updated Provider Description";
-      Optional<Provider> updatedProvider = providerService.updateProvider(providerId,
-          getTestProviderInputModelBuilder(TEST_EVENT_PROVIDER_LABEL)
+      Provider updatedProvider = createOrUpdateProvider(getTestProviderInputModelBuilder(TEST_EVENT_PROVIDER_LABEL)
```
Comment (abhupadh) [inline events_test/src/test/java/com/adobe/aio/event/management/ProviderServiceIntegrationTest.java:114]: here

File: events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java (hunk 4 above)
Comment (abhupadh) [reply events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java:63]: Well,I have done this [here](https://github.com/adobe/aio-lib-java/pull/200/commits/8f19718e7ef1057a50dda92a76e061fb4fc17bff#r1450096984) in the `ProviderServiceIntegrationTest`, is there any other place I need to do that you meant?

**Review by francoisledroff — APPROVED** [review]

File: events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java (hunk 4 above)
Comment (francoisledroff) [reply events_test/src/test/java/com/adobe/aio/event/management/RegistrationServiceIntegrationTest.java:63]: oh I did not catch this, thanks !

**Pull request merged**